src/rust/scheduler/scheduler.rs

```diff
@@ -84,157 +84,175 @@ pub struct Scheduler {
+                "remove(): name={:?}, id={:?}, pin_slab_index={:?}",
                 task.get_name(),
                 task_id,
-                index
+                pin_slab_index
             );
             Some(task)
         } else {
-            warn!("Unable to unpin and remove: id={:?}, index={:?}", task_id, index);
+            warn!("Unable to unpin and remove: id={:?}, pin_slab_index={:?}", task_id, pin_slab_index);
             None
         }
     }
 
     /// Given a task id return a handle to the task.
     pub fn from_task_id(&self, task_id: u64) -> Option<TaskHandle> {
-        let pages: Ref<Vec<WakerPageRef>> = self.pages.borrow();
-        let index: usize = match self.task_ids.borrow().get(&task_id) {
-            Some(index) => *index,
+        let page_refs: Ref<Vec<WakerPageRef>> = self.page_refs.borrow();
+        let pin_slab_index: usize = match self.task_ids.borrow().get(&task_id) {
+            Some(pin_slab_index) => *pin_slab_index,
             None => return None,
         };
-        self.tasks.borrow().get(index)?;
-        let page: &WakerPageRef = {
-            let (pages_ix, _) = self.get_page_indexes(index);
-            &pages[pages_ix]
+        self.tasks.borrow().get(pin_slab_index)?;
+        let page_ref: &WakerPageRef = {
+            let (page_index, _) = self.get_page_index_and_offset(pin_slab_index);
+            &page_refs[page_index]
         };
-        let handle: TaskHandle = TaskHandle::new(task_id, index, page.clone());
+        let handle: TaskHandle = TaskHandle::new(task_id, pin_slab_index, page_ref.clone());
         Some(handle)
     }
 
     /// Insert a new task into our scheduler returning a handle corresponding to it.
     pub fn insert<F: Task>(&self, future: F) -> Option<TaskHandle> {
-        let mut pages: RefMut<Vec<WakerPageRef>> = self.pages.borrow_mut();
-        let mut id_gen: RefMut<SmallRng> = self.id_gen.borrow_mut();
+        self.panic_if_too_many_tasks();
+
         let task_name: String = future.get_name();
-        // Allocate an offset into the slab and a token for identifying the task.
-        let index: usize = self.tasks.borrow_mut().insert(Box::new(future))?;
+        // The pin slab index can be reverse-computed in a page index and an offset within the page.
+        let pin_slab_index: usize = self.tasks.borrow_mut().insert(Box::new(future))?;
 
-        // Generate a new id. If the id is currently in use, keep generating until we find an unused id.
-        let mut task_ids: RefMut<HashMap<u64, usize>> = self.task_ids.borrow_mut();
-        // If the address space for task ids is close to half full, it will become increasingly difficult to avoid
-        // collisions, so we cap the number of tasks at 16,000.
-        if task_ids.len() > MAX_NUM_TASKS {
-            panic!("Too many concurrent tasks");
-        }
-        let task_id: u64 = 'get_id: {
+        self.add_new_pages_to_accommodate_index_gap(pin_slab_index);
```

Comment:
maybe name this: allocate_new_pages_up_to_waker_index or something like that

Reply:
Just updated it to match up to that meaning. Let me know if that looks ok.